### ISSUE TITLE
DGESVD failure on some matrices of decaying exponentials

### ISSUE THREAD

**Issue by jasonkaye** [xianyi/OpenBLAS#4583]
I am encountering a strange failure using DGESVD on a particular matrix. I am using OpenBLAS 0.3.26, installed via homebrew, on an Apple M2 with MacOS 14.2.1. I have attempted to boil it down to a minimal example. I fill in a matrix with some decaying exponentials, and try to take its SVD.

```
#include <lapack.h>
#include <iostream>
#include <vector>

int main() {

  // matrix dimensions
  int m = 400;
  int n = 20;
  int size = m * n;

  // allocate memory
  std::vector<double> A(m * n, 0.0);
  std::vector<double> s(n, 0.0);
  std::vector<double> U(m * m, 0.0);
  std::vector<double> Vt(n * n, 0.0);

  // fill matrix A
  double c = 800;
  for (int j = 0; j < n; ++j) {
    for (int i = 0; i < m; ++i) {
      std::size_t col_idx = n - j - 1;
      std::size_t row_idx = m - i - 1;
      std::size_t idx = col_idx * m + row_idx;
      A[idx] = exp(-i * j * c / size);
    }
  }

  // openblas call to dgesvd
  char jobu = 'A';
  char jobvt = 'A';
  int lda = m;
  int ldu = m;
  int ldvt = n;
  double tmp_work;
  int lwork = -1;
  int info;
  LAPACK_dgesvd(&jobu, &jobvt, &m, &n, A.data(), &lda, s.data(), U.data(), &ldu, Vt.data(), &ldvt, &tmp_work, &lwork, &info);
  lwork = static_cast<int>(tmp_work);
  std::vector<double> work(lwork, 0.0);
  LAPACK_dgesvd(&jobu, &jobvt, &m, &n, A.data(), &lda, s.data(), U.data(), &ldu, Vt.data(), &ldvt, work.data(), &lwork, &info);

  // output results
  std::cout << "info = " << info << std::endl;
} 
```

I am compiling the program with Clang 17.0.6. The result is info = 19. The SVD works perfectly fine (the error goes away) when I

- Use OpenBLAS 0.3.25
- Use OpenBLAS 0.3.26 in Linux
- Use OpenBLAS 0.3.26 on a Mac with an Intel chip
- Dump out the matrix and compute the SVD using MATLAB
- Decrease m, n, or c significantly
- More bizarrely, replace n - j - 1 and m - i - 1 with j and i, respectively, which simply reverses the order of rows and columns, or
- Add random noise on the level of 10^{-15} to the matrix.

Compiling OpenBLAS 0.3.26 from source does not fix the problem, nor does switching to a different M1/M2 Mac.

The critical value of c is suspiciously close to the point at which the exponential starts to underflow for some of the values; this should not be a problem, but perhaps it has something to do with it.

**Comment by martin-frbg** [comment]
Hmm, strange - DGESVD itself has not changed between 0.3.25 and 0.3.26, nor has the DGEMM kernel used on Apple M. On the other hand, what _has_ changed is that multi-cpu (DYNAMIC_ARCH) builds of OpenBLAS (which is probably what you get from Homebrew) will now contain dedicated support for the M cpu where they used to default to a more generic ARMV8 target earlier. (This should be a moot point if you built from source without specifying the DYNAMIC_ARCH option, and otherwise it should also not directly affect DGEMM but rather DAXPY and DDOT - this would have to be some very indirect effect on intermediate result precision. Do you get the "old"  result when you `export OPENBLAS_CORETYPE=ARMV8` in the environment before running your code ?)

**Comment by jasonkaye** [comment]
Thanks for the suggestion @martin-frbg, this indeed works around the issue. Is there a way to make that switch to ARMV8 at runtime through an openblas library call? This would be a more viable temporary workaround for the library that I am maintaining.

**Comment by martin-frbg** [comment]
Unfortunately not - but I hope to have this fixed in 0.3.27, which should be released in a couple of days anyway.

This appears to be due to a loss of precision in the DNRM2 kernel used on Apple M cpus, so actually caused by making the "VORTEX" target available in DYNAMIC_ARCH builds in 0.3.26 - while the problem in that particular kernel is older than 0.3.23 (still bisecting to find the exact commit that caused it, probably related to mitigation attempts for wrong behaviour in other corner cases)

**Comment by jasonkaye** [comment]
Great, thanks for the help. I will try this in the new release.

**Comment by jasonkaye** [comment]
While the problem above is indeed fixed in OpenBLAS 0.3.27, it persists if one changes the double matrices to complex double, and uses LAPACK_zgesvd instead of LAPACK_dgesvd. However, the modification below fixes the issue in that case as well:

`diff --git a/kernel/arm64/KERNEL.NEOVERSEN1 b/kernel/arm64/KERNEL.NEOVERSEN1
index 9a5938459b..9fe981c585 100644
--- a/kernel/arm64/KERNEL.NEOVERSEN1
+++ b/kernel/arm64/KERNEL.NEOVERSEN1
@@ -91,8 +91,8 @@ IDAMAXKERNEL   = iamax_thunderx2t99.c
 ICAMAXKERNEL   = izamax_thunderx2t99.c
 IZAMAXKERNEL   = izamax_thunderx2t99.c
 
-SNRM2KERNEL    = scnrm2_thunderx2t99.c
-DNRM2KERNEL    = dznrm2_thunderx2t99.c
+SNRM2KERNEL    = nrm2.S
+DNRM2KERNEL    = nrm2.S
 CNRM2KERNEL    = scnrm2_thunderx2t99.c
 ZNRM2KERNEL    = dznrm2_thunderx2t99.c`

Does it make sense to make this change in OpenBLAS?
